File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/ShoppingListItemsAdapter.java
package za.co.woolworths.financial.services.android.ui.adapters;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

import android.graphics.Color;
import android.graphics.PorterDuff;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.awfs.coordination.R;
import com.daimajia.swipe.SwipeLayout;
import com.daimajia.swipe.adapters.RecyclerSwipeAdapter;

import java.util.ArrayList;
import java.util.List;

import za.co.woolworths.financial.services.android.models.dto.OtherSkus;
import za.co.woolworths.financial.services.android.models.dto.ProductList;
import za.co.woolworths.financial.services.android.models.dto.ShoppingDeliveryLocation;
import za.co.woolworths.financial.services.android.models.dto.ShoppingListItem;
import za.co.woolworths.financial.services.android.ui.fragments.shoppinglist.listitems.ShoppingListItemsNavigator;
import za.co.woolworths.financial.services.android.ui.views.WTextView;
import za.co.woolworths.financial.services.android.ui.views.WrapContentDraweeView;
import za.co.woolworths.financial.services.android.util.CurrencyFormatter;
import za.co.woolworths.financial.services.android.util.KotlinUtils;
import za.co.woolworths.financial.services.android.util.Utils;
import za.co.woolworths.financial.services.android.util.analytics.FirebaseManager;
import za.co.woolworths.financial.services.android.util.wenum.Delivery;

/**
 * Created by W7099877 on 2018/03/09.
 */

public class ShoppingListItemsAdapter extends RecyclerSwipeAdapter<RecyclerView.ViewHolder> {

	private final int ITEM_VIEW_TYPE_HEADER = 0;
	private final int ITEM_VIEW_TYPE_BASIC = 1;

	private ArrayList<ShoppingListItem> mShoppingListItem;
	private final ShoppingListItemsNavigator navigator;
	private boolean mAdapterIsClickable;


	public ShoppingListItemsAdapter(ArrayList<ShoppingListItem> shoppingListItems, ShoppingListItemsNavigator navigator) {
		this.mShoppingListItem = shoppingListItems;
		this.navigator = navigator;

	}

	@NonNull
	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

		RecyclerView.ViewHolder vh;
		switch (viewType) {
			case ITEM_VIEW_TYPE_HEADER:
				vh = getHeaderViewHolder(parent);
				break;

			case ITEM_VIEW_TYPE_BASIC:
			default:
				vh = getSimpleViewHolder(parent);
				break;

		}

		return vh;
	}

	private RecyclerView.ViewHolder getHeaderViewHolder(ViewGroup parent) {
		return new HeaderViewHolder(LayoutInflater.from(parent.getContext()).inflate(
				R.layout.search_add_to_list_layout, parent, false));
	}

	private RecyclerView.ViewHolder getSimpleViewHolder(ViewGroup parent) {
		return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(
				R.layout.shopping_list_commerce_item, parent, false));
	}

	@Override
	public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, final int position) {
		switch (getItemViewType(position)) {
			case ITEM_VIEW_TYPE_HEADER:
				HeaderViewHolder headerViewHolder = (HeaderViewHolder) viewHolder;
				headerViewHolder.tvSearchText.setOnClickListener(view -> navigator.onShoppingSearchClick());
				break;

			case ITEM_VIEW_TYPE_BASIC:
				final ViewHolder holder = (ViewHolder) viewHolder;
				final ShoppingListItem shoppingListItem = getItem(position);
				if (shoppingListItem == null) return;
				if (shoppingListItem.displayName == null) return;
				holder.cartProductImage.setImageURI(shoppingListItem.externalImageRefV2);
				holder.productName.setText(shoppingListItem.displayName);
				holder.tvQuantity.setText(String.valueOf(shoppingListItem.userQuantity));
				holder.price.setText(CurrencyFormatter.Companion.formatAmountToRandAndCentWithSpace(shoppingListItem.price));
				holder.cbxSelectShoppingListItem.setChecked(shoppingListItem.isSelected);
				holder.delete.setVisibility(VISIBLE);
				holder.progressBar.setVisibility(View.INVISIBLE);
				holder.tvProductAvailability.setVisibility(GONE);
				holder.swipeLayout.setTopSwipeEnabled(false);

				holder.swipeLayout.addDrag(SwipeLayout.DragEdge.Right, holder.swipeRight);

				// Set Color and Size START
				String sizeColor = shoppingListItem.color;
				if (sizeColor == null)
					sizeColor = "";
				if (sizeColor.isEmpty() && !shoppingListItem.size.isEmpty() && !shoppingListItem.size.equalsIgnoreCase("NO SZ"))
					sizeColor = shoppingListItem.size;
				else if (!sizeColor.isEmpty() && !shoppingListItem.size.isEmpty() && !shoppingListItem.size.equalsIgnoreCase("NO SZ"))
					sizeColor = sizeColor + ", " + shoppingListItem.size;

				holder.tvColorSize.setText(sizeColor);
				holder.tvColorSize.setVisibility(VISIBLE);
				/****
				 * shoppingListItem.userShouldSetSuburb - is set to true when user did not cbxSelectShoppingListItem any suburb
				 */

				if (userShouldSetSuburb()) {
					holder.llQuantity.setVisibility(VISIBLE);
					holder.llQuantity.setAlpha(1.0f);
					holder.cbxSelectShoppingListItem.setEnabled(true);
					adapterClickable(true);
					holder.cbxSelectShoppingListItem.setAlpha(1.0f);
					holder.llQuantity.setEnabled(true);
				} else {
					boolean productInStock = shoppingListItem.quantityInStock != 0;
					holder.llQuantity.setAlpha(productInStock ? 1.0f : 0.5f);
					holder.tvQuantity.setAlpha(productInStock ? 1.0f : 0.5f);
					holder.cbxSelectShoppingListItem.setEnabled(productInStock);
					holder.pbQuantityLoader.setVisibility(VISIBLE);
					holder.cbxSelectShoppingListItem.setEnabled(false);
					holder.pbQuantityLoader.getIndeterminateDrawable().setColorFilter(Color.BLACK, PorterDuff.Mode.MULTIPLY);
					if (shoppingListItem.inventoryCallCompleted) {
						int inventoryQueryStatus = shoppingListItem.quantityInStock;
						holder.pbQuantityLoader.setVisibility(GONE);
						holder.cbxSelectShoppingListItem.setEnabled(true);
						if (inventoryQueryStatus == -1) {
							holder.llQuantity.setVisibility(GONE);
							holder.cbxSelectShoppingListItem.setVisibility(GONE);
							holder.tvColorSize.setVisibility(GONE);
							holder.tvProductAvailability.setVisibility(VISIBLE);
							holder.price.setAlpha(0f);
							holder.price.setVisibility(GONE);
						} else {
							holder.llQuantity.setVisibility((shoppingListItem.quantityInStock == 0 || !shoppingListItem.isSelected) ? GONE : VISIBLE);
							holder.minusDeleteCountImage.setImageResource(
									shoppingListItem.userQuantity == 1
											&& shoppingListItem.isSelected ?
											R.drawable.delete_24 : R.drawable.ic_minus_black
							);
							int padding = (int) holder.minusDeleteCountImage.getContext()
									.getResources().getDimension(shoppingListItem.userQuantity == 1
											&& shoppingListItem.isSelected ? R.dimen.seven_dp : R.dimen.ten_dp);
							holder.minusDeleteCountImage.setPadding(padding, padding, padding, padding);
							holder.minusDeleteCountImage.setVisibility(shoppingListItem.isSelected ? VISIBLE : GONE);
							holder.addCountImage.setVisibility(
									(shoppingListItem.quantityInStock == 1 ||
											shoppingListItem.userQuantity == shoppingListItem.quantityInStock)
											? GONE : VISIBLE
							);
							holder.tvProductAvailability.setVisibility((shoppingListItem.quantityInStock == 0) ? VISIBLE : GONE);
							holder.cbxSelectShoppingListItem.setVisibility((shoppingListItem.quantityInStock == 0) ? GONE : VISIBLE);
							holder.price.setVisibility((shoppingListItem.quantityInStock == 0) ? GONE : VISIBLE);
							holder.price.setAlpha(1f);
							holder.tvColorSize.setVisibility(VISIBLE);
						}

						int msg = getUnavailableMessage(shoppingListItem.unavailable);
						Utils.setBackgroundColor(
								holder.tvProductAvailability,
								R.drawable.delivery_round_btn_black,
								msg
						);
					} else {
						holder.llQuantity.setVisibility(GONE);
						holder.tvProductAvailability.setVisibility(GONE);
						holder.cbxSelectShoppingListItem.setEnabled(false);
					}
				}

				if (!userShouldSetSuburb())
					if (!shoppingListItem.inventoryCallCompleted) {
						holder.llQuantity.setAlpha(0.5f);
						holder.tvQuantity.setAlpha(0.5f);
					}

				holder.minusDeleteCountImage.setOnClickListener(view -> {
					ShoppingListItem listItem = getItem(position);
					if (enableClickEvent(listItem))
						return;
					if (!mAdapterIsClickable) return;
					if (userShouldSetSuburb()) {
						navigator.openSetSuburbProcess(listItem);
						return;
					}
					if (listItem.quantityInStock == 0) return;
					// One item selected by user in add to list
					if (listItem.userQuantity == 1) {
						deleteItemFromList(listItem, position);
					} else if (listItem.userQuantity > 1) {
						listItem.userQuantity -= 1;
						mShoppingListItem.set((position-1), listItem);
						navigator.onSubstractListItemCount(listItem);
						notifyItemChanged(position, listItem);
					}
				});

				holder.addCountImage.setOnClickListener(view -> {
					ShoppingListItem listItem = getItem(position);
					if (enableClickEvent(listItem))
						return;
					if (!mAdapterIsClickable) return;
					if (userShouldSetSuburb()) {
						navigator.openSetSuburbProcess(listItem);
						return;
					}
					if (listItem.quantityInStock == 0) return;
					// One item selected by user in add to list
					if (listItem.userQuantity < listItem.quantityInStock) {
						listItem.userQuantity += 1;
						mShoppingListItem.set((position - 1), listItem);
						navigator.onAddListItemCount(listItem);
						notifyItemChanged(position, listItem);
					}
				});

                holder.llItemContainer.setOnClickListener(view -> {
                    ShoppingListItem shoppingListItem14 = getItem(position);
                    if (shoppingListItem14.unavailable)
                        navigator.showListBlackToolTip();
                });

				holder.cartProductImage.setOnClickListener(view -> {
					if (!mAdapterIsClickable) return;
					ShoppingListItem shoppingListItem12 = getItem(position);
					ProductList productList = createProductList(shoppingListItem12);
					navigator.openProductDetailFragment(shoppingListItem12.displayName, productList);
				});

				// Set Color and Size END
				holder.cbxSelectShoppingListItem.setOnClickListener(view -> {
					ShoppingListItem item = getItem(position);
					if (enableClickEvent(item))
						return;
					if (!mAdapterIsClickable) return;
					if (!item.isSelected) {
						if (userShouldSetSuburb()) {
							item.isSelected = false;
							notifyItemChanged(position, mShoppingListItem.size());
							navigator.openSetSuburbProcess(item);
							return;
						}
					}
					if (item.quantityInStock == 0) return;
                    /*
                     1. By default quantity will be ZERO.
                     2. On Selection it will change to ONE.
                     */
					item.userQuantity = Math.max(item.userQuantity, 1);
					item.isSelected = !item.isSelected;
					mShoppingListItem.set((position - 1), item);
					navigator.onItemSelectionChange(item.isSelected);
					notifyItemChanged(position, item);
				});

				holder.delete.setOnClickListener(view -> {
					if (!mAdapterIsClickable) return;
					navigator.onItemDeleteClick(getItem(position).Id, getItem(position).productId, getItem(position).catalogRefId, true);
				});
				mItemManger.bindView(holder.itemView, position);
				break;
			default:
				break;
		}
	}

    private int getUnavailableMessage(boolean isUnavailable) {
        int msg = isUnavailable ? R.string.unavailable : R.string.out_of_stock;
        if (isUnavailable) {
            Delivery type = KotlinUtils.Companion.getPreferredDeliveryType();
            if (type == null) {
                return msg;
            }
            switch (type) {
                case DASH:
                    msg = R.string.unavailable_with_dash;
                    break;
                case CNC:
                    msg = R.string.unavailable_with_collection;
                    break;
            }
        }
        return msg;
    }

	private void deleteItemFromList(ShoppingListItem shoppingListItem, int adapterPosition) {
		if (mShoppingListItem == null || mShoppingListItem.size() <= 0) {
			return;
		}
		navigator.onItemDeleteClick(shoppingListItem.Id, shoppingListItem.productId, shoppingListItem.catalogRefId, true);
	}

	private boolean enableClickEvent(ShoppingListItem shoppingListItem) {
		if (!userShouldSetSuburb())
			return shoppingListItem.quantityInStock == -1;
		return false;
	}

	@NonNull
	private ProductList createProductList(ShoppingListItem shoppingListItem) {
		ProductList productList = new ProductList();
		productList.productId = shoppingListItem.productId;
		productList.productName = shoppingListItem.displayName;
		productList.fromPrice = Float.valueOf((TextUtils.isEmpty(shoppingListItem.price) ? "0.0" : shoppingListItem.price));
		productList.sku = shoppingListItem.catalogRefId;
		productList.externalImageRefV2 = shoppingListItem.externalImageRefV2;
		OtherSkus otherSku = new OtherSkus();
		otherSku.price = String.valueOf(shoppingListItem.price);
		otherSku.size = "";
		List<OtherSkus> otherSkuList = new ArrayList<>();
		productList.otherSkus = new ArrayList<>();
		otherSkuList.add(otherSku);
		productList.otherSkus = otherSkuList;
		return productList;
	}

	@Override
	public int getItemCount() {
		return mShoppingListItem.size() + 1;
	}

	public Integer getAddedItemsCount() {
		int count = 0;
		if (mShoppingListItem == null || mShoppingListItem.isEmpty()) return count;
		for (ShoppingListItem item : mShoppingListItem) {
			if (item.isSelected) {
				count += item.userQuantity;
			}
		}
		return count;
	}

	@Override
	public int getSwipeLayoutResourceId(int position) {
		return R.id.swipe;
	}

	class ViewHolder extends RecyclerView.ViewHolder {
		private final WTextView productName;
		private final WTextView tvColorSize;
		private final WTextView tvQuantity;
		private final WTextView price;
		private final CheckBox cbxSelectShoppingListItem;
		private final WrapContentDraweeView cartProductImage;
		private final WTextView delete;
		private final ProgressBar progressBar;
		private final ProgressBar pbQuantityLoader;
		private final LinearLayout llQuantity;
		private final ConstraintLayout llItemContainer;
		private final RelativeLayout rlQuantitySelector;
		private final SwipeLayout swipeLayout;
		private final TextView tvProductAvailability;
		private final RelativeLayout swipeRight;
		private final ImageView minusDeleteCountImage, addCountImage;

		public ViewHolder(View itemView) {
			super(itemView);
			productName = itemView.findViewById(R.id.tvTitle);
			tvQuantity = itemView.findViewById(R.id.tvQuantity);
			price = itemView.findViewById(R.id.tvPrice);
			cbxSelectShoppingListItem = itemView.findViewById(R.id.cbShoppingList);
			pbQuantityLoader = itemView.findViewById(R.id.pbQuantityLoader);
			cartProductImage = itemView.findViewById(R.id.cartProductImage);
			tvProductAvailability = itemView.findViewById(R.id.tvProductAvailability);
			delete = itemView.findViewById(R.id.tvDelete);
			progressBar = itemView.findViewById(R.id.pbDeleteIndicator);
			tvColorSize = itemView.findViewById(R.id.tvColorSize);
			llQuantity = itemView.findViewById(R.id.llQuantity);
			llItemContainer = itemView.findViewById(R.id.llItemContainer);
			minusDeleteCountImage = itemView.findViewById(R.id.minusDeleteCountImage);
			addCountImage = itemView.findViewById(R.id.addCountImage);
			rlQuantitySelector = itemView.findViewById(R.id.rlQuantitySelector);
			swipeLayout = itemView.findViewById(R.id.swipe);
			swipeRight = itemView.findViewById(R.id.swipeRight);
		}
	}

	class HeaderViewHolder extends RecyclerView.ViewHolder {
		private final WTextView tvSearchText;

		public HeaderViewHolder(View itemView) {
			super(itemView);
			tvSearchText = itemView.findViewById(R.id.textProductSearch);
		}
	}

	/***
	 * Update old list with new list before refreshing the adapter
	 */
	public void updateList(ArrayList<ShoppingListItem> updatedListItems) {
		if (updatedListItems == null) return;
		if (mShoppingListItem == null) return;
		try {
			for (ShoppingListItem shoppinglistItem : mShoppingListItem) {
				for (ShoppingListItem updatedList : updatedListItems) {
					if (shoppinglistItem.catalogRefId.equalsIgnoreCase(updatedList.catalogRefId)) {
						int userQuantity = (updatedList.unavailable || updatedList.quantityInStock == 0)
								? 0 : shoppinglistItem.userQuantity;
						userQuantity = userQuantity > updatedList.quantityInStock ? 1 : userQuantity;
						updatedList.userQuantity =  userQuantity;

						updatedList.isSelected = !updatedList.unavailable &&
								updatedList.quantityInStock != 0
								&& shoppinglistItem.isSelected;
					}
				}
			}
			this.mShoppingListItem = updatedListItems;
			navigator.onItemSelectionChange(false); // default value
			notifyDataSetChanged();
			closeAllItems();
		} catch (IllegalArgumentException ex) {
			FirebaseManager.logException(ex);
		}
	}

	@Override
	public int getItemViewType(int position) {
		return isPositionHeader(position) ? ITEM_VIEW_TYPE_HEADER : ITEM_VIEW_TYPE_BASIC;
	}

	public void adapterClickable(boolean clickable) {
		this.mAdapterIsClickable = clickable;
	}

	private boolean isPositionHeader(int position) {
		return position == 0;
	}

	private ShoppingListItem getItem(int position) {
		return mShoppingListItem.get(position - 1);
	}

	public ArrayList<ShoppingListItem> getShoppingListItems() {
		return mShoppingListItem;
	}

	public boolean userShouldSetSuburb() {
		return Utils.getPreferredDeliveryLocation() == null;
	}

	public void resetSelection() {
		for (ShoppingListItem shoppinglistItem : mShoppingListItem) {
			shoppinglistItem.userQuantity = 0;
			shoppinglistItem.isSelected = false;
		}
		notifyDataSetChanged();
		navigator.onItemSelectionChange(false);
	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/ShoppingListItemsAdapter.kt
package za.co.woolworths.financial.services.android.ui.adapters

import android.content.Context
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.awfs.coordination.R
import com.awfs.coordination.databinding.ShoppingListCommerceItemBinding
import com.daimajia.swipe.SwipeLayout
import com.daimajia.swipe.adapters.RecyclerSwipeAdapter
import za.co.woolworths.financial.services.android.models.dto.OtherSkus
import za.co.woolworths.financial.services.android.models.dto.ProductList
import za.co.woolworths.financial.services.android.models.dto.ShoppingListItem
import za.co.woolworths.financial.services.android.shoppinglist.component.MyListFlowType
import za.co.woolworths.financial.services.android.ui.fragments.shoppinglist.listitems.ProductAvailability
import za.co.woolworths.financial.services.android.ui.fragments.shoppinglist.listitems.ShoppingListItemsNavigator
import za.co.woolworths.financial.services.android.util.AppConstant.Companion.CONST_NO_SIZE
import za.co.woolworths.financial.services.android.util.CurrencyFormatter.Companion.formatAmountToRandAndCentWithSpace
import za.co.woolworths.financial.services.android.util.KotlinUtils.Companion.getPreferredDeliveryType
import za.co.woolworths.financial.services.android.util.Utils
import za.co.woolworths.financial.services.android.util.Utils.getString
import za.co.woolworths.financial.services.android.util.wenum.Delivery
import java.util.*

/**
 * Created by W7099877 on 2018/03/09.
 */
class ShoppingListItemsAdapter(
    var shoppingListItems: ArrayList<ShoppingListItem>?,
    private val navigator: ShoppingListItemsNavigator
) : RecyclerSwipeAdapter<RecyclerView.ViewHolder>() {

    companion object {

        private const val ITEM_VIEW_TYPE_HEADER = 0
        private const val ITEM_VIEW_TYPE_BASIC = 1
    }

    private var type = getPreferredDeliveryType()
    private var userShouldSetSuburb = userShouldSetSuburb()
    private var mAdapterIsClickable = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
        when (viewType) {
            ITEM_VIEW_TYPE_HEADER -> getHeaderViewHolder(parent)
            ITEM_VIEW_TYPE_BASIC -> getSimpleViewHolder(parent)
            else -> getSimpleViewHolder(parent)
        }

    private fun getHeaderViewHolder(parent: ViewGroup): RecyclerView.ViewHolder = HeaderViewHolder(
        LayoutInflater.from(parent.context).inflate(
            R.layout.search_add_to_list_layout, parent, false
        )
    )

    private fun getSimpleViewHolder(parent: ViewGroup): RecyclerView.ViewHolder {
        return ShoppingListItemViewHolder(
            ShoppingListCommerceItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            ITEM_VIEW_TYPE_HEADER -> {
                val headerViewHolder = viewHolder as HeaderViewHolder
                headerViewHolder.tvSearchText.setOnClickListener { navigator.onShoppingSearchClick() }
            }

            ITEM_VIEW_TYPE_BASIC -> {
                val holder = viewHolder as? ShoppingListItemViewHolder ?: return
                val shoppingListItem = getItem(position) ?: return

                holder.bind(shoppingListItem)

                when (shoppingListItem.availability) {
                    ProductAvailability.UNAVAILABLE.value -> holder.bindUnavailableProduct()
                    ProductAvailability.OUT_OF_STOCK.value -> holder.bindOutOfStockProduct(
                        shoppingListItem
                    )

                    else -> holder.bindAvailableProduct(shoppingListItem)
                }
            }
        }
    }

    private fun getColorAndSize(color: String?, size: String?): String = buildString {
        append(color ?: "")

        if (size.isNullOrEmpty() || CONST_NO_SIZE.equals(size, ignoreCase = true)) {
            return@buildString
        }
        append(if (!color.isNullOrEmpty()) ", " else "")
        append(size)
    }

    private fun getUnavailableMsgByDeliveryType(context: Context): String {
        val message = getString(context, R.string.unavailable)
        type ?: return message
        return getString(
            context,
            when (type) {
                Delivery.DASH -> R.string.unavailable_with_dash
                Delivery.CNC -> R.string.unavailable_with_collection
                else -> R.string.unavailable
            }
        )
    }

    private fun deleteItemFromList(shoppingListItem: ShoppingListItem, adapterPosition: Int) {
        navigator.onItemDeleteClick(shoppingListItem)
    }

    private fun enableClickEvent(shoppingListItem: ShoppingListItem): Boolean {
        return if (!userShouldSetSuburb()) shoppingListItem.quantityInStock == -1 else false
    }

    private fun createProductList(shoppingListItem: ShoppingListItem): ProductList {
        val productList = ProductList().apply {
            productId = shoppingListItem.productId
            productName = shoppingListItem.displayName
            sku = shoppingListItem.catalogRefId
            externalImageRefV2 = shoppingListItem.externalImageRefV2

            fromPrice = if (shoppingListItem.price.isNullOrEmpty()) "0.0".toFloat()
            else shoppingListItem.price.toFloat()

            val otherSku = OtherSkus().also {
                it.price = shoppingListItem.price.toString()
                it.size = ""
            }
            val otherSkuList: MutableList<OtherSkus> = ArrayList(0)
            otherSkuList.add(otherSku)
            otherSkus = otherSkuList
        }
        return productList
    }

    override fun getItemCount(): Int {
        return shoppingListItems?.size?.plus(1) ?: 1
    }

    val addedItemsCount: Int
        get() {
            var count = 0
            shoppingListItems ?: return count
            shoppingListItems?.map { item ->
                if (item.isSelected)
                    count += item.userQuantity
            }
            return count
        }

    override fun getSwipeLayoutResourceId(position: Int): Int {
        return R.id.swipe
    }

    internal inner class ShoppingListItemViewHolder(val itemBinding: ShoppingListCommerceItemBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        fun bind(shoppingListItem: ShoppingListItem) {
            itemBinding?.apply {

                cartProductImage.setImageURI(shoppingListItem.externalImageRefV2)
                tvTitle.text = (shoppingListItem.displayName)
                tvPrice.setText(formatAmountToRandAndCentWithSpace(shoppingListItem.price))
                // Set Color and Size START
                tvColorSize.text =
                    getColorAndSize(shoppingListItem.color, shoppingListItem.size)
                //Swipe Layout
                mItemManger.bindView(root, position)
                swipe.isTopSwipeEnabled = false
                swipe.addDrag(SwipeLayout.DragEdge.Right, swipeRight)

                tvProductAvailability.visibility = GONE
                cbShoppingList.visibility = GONE
                cbShoppingList.isEnabled = false
                llQuantity.visibility = GONE
                rlAddToCart.visibility = if (shoppingListItem.quantityInStock > 0) VISIBLE else GONE
                pbAddIndicator.visibility =
                    if (shoppingListItem.isAddToCartInProgress) VISIBLE else GONE

                // Product Image
                cartProductImage.setOnClickListener {
                    if (!mAdapterIsClickable) return@setOnClickListener
                    val listItem = getItem(position) ?: return@setOnClickListener
                    val productList = createProductList(listItem)
                    navigator.openProductDetailFragment(listItem.displayName, productList)
                }

                // Swipe delete click
                tvDelete.setOnClickListener {
                    if (!mAdapterIsClickable) return@setOnClickListener
                    val item = getItem(position) ?: return@setOnClickListener
                    navigator.onItemDeleteClick(item)
                }
                // Swipe to add click
                tvAddItem.setOnClickListener {
                    if (!mAdapterIsClickable) return@setOnClickListener
                    adapterClickable(false)
                    val item = getItem(position) ?: return@setOnClickListener
                    item.userQuantity = 1
                    item.isSelected = true
                    item.isAddToCartInProgress = true
                    shoppingListItems?.set(position - 1, item)
                    notifyItemChanged(position)
                    pbAddIndicator.visibility = VISIBLE
                    navigator.onItemAddClick(item)
                }
            }
        }

        fun bindUnavailableProduct() {
            itemBinding?.apply {
                iconKebab.visibility = GONE
                adapterClickable(true)
                val msg = getUnavailableMsgByDeliveryType(itemBinding.root.context)
                tvProductAvailability.text = msg
                tvProductAvailability.visibility = VISIBLE
            }
        }

        fun bindOutOfStockProduct(shoppingListItem: ShoppingListItem) {
            itemBinding?.apply {
                iconKebab.visibility = VISIBLE
                adapterClickable(true)
                tvProductAvailability.text =
                    itemBinding.root.context.getString(R.string.out_of_stock)
                tvProductAvailability.visibility = VISIBLE
                iconKebab.setOnClickListener {
                    navigator.naviagteToMoreOptionDialog(shoppingListItem)
                }
            }
        }

        fun bindAvailableProduct(shoppingListItem: ShoppingListItem) {

            itemBinding?.apply {

                if (!shoppingListItem.inventoryCallCompleted) {
                    pbQuantityLoader.visibility = VISIBLE
                    return
                }

                //Inventory progress bar
                pbQuantityLoader.visibility = GONE

                if (shoppingListItem.quantityInStock <= 0) {
                    return
                }

                //enable check box
                cbShoppingList.visibility = VISIBLE
                cbShoppingList.isEnabled = true

                /*
                * shoppingListItem.userShouldSetSuburb - is set to true when user did not cbxSelectShoppingListItem any suburb
                */
                if (userShouldSetSuburb) {
                    adapterClickable(true)
                    return
                }

                cbShoppingList.isChecked = shoppingListItem.isSelected
                tvQuantity.setText(shoppingListItem.userQuantity.toString())
                // Delete button
                val padding = root.context.resources.getDimension(
                    if (shoppingListItem.userQuantity == 1 && shoppingListItem.isSelected)
                        R.dimen.seven_dp else R.dimen.ten_dp
                ).toInt()
                minusDeleteCountImage.setPadding(padding, padding, padding, padding)
                if (shoppingListItem.userQuantity == 1) {
                    if (MyListFlowType.getFlowType() == MyListFlowType.FlowTypeViewOnly) {
                        minusDeleteCountImage.setImageResource(R.drawable.disabled_delete_icon)
                        minusDeleteCountImage.isEnabled = false
                    } else {
                        minusDeleteCountImage.setImageResource(R.drawable.delete_24)
                        minusDeleteCountImage.isEnabled = true
                    }
                } else {
                    minusDeleteCountImage.setImageResource(R.drawable.ic_minus_black)
                    minusDeleteCountImage.isEnabled = true
                }

                // Add button
                addCountImage.visibility =
                    if (shoppingListItem.quantityInStock == 1 ||
                        shoppingListItem.userQuantity == shoppingListItem.quantityInStock
                    ) GONE else VISIBLE

                llQuantity.visibility = if (shoppingListItem.isSelected) VISIBLE else GONE

                cbShoppingList.setOnClickListener {

                    val item = getItem(bindingAdapterPosition) ?: return@setOnClickListener
                    if (enableClickEvent(item) || !mAdapterIsClickable) return@setOnClickListener

                    if (swipe.isSwipeEnabled){
                        swipe.close()
                    }

                    if (!item.isSelected && userShouldSetSuburb()) {
                        item.isSelected = false
                        notifyItemChanged(position, shoppingListItems?.size)
                        navigator.openSetSuburbProcess(item)
                        return@setOnClickListener
                    }
                    if (item.quantityInStock == 0) return@setOnClickListener
                    /*
                     1. By default quantity will be ZERO.
                     2. On Selection it will change to ONE.
                     */
                    item.userQuantity = item.userQuantity.coerceAtLeast(1)
                    item.isSelected = !item.isSelected
                    shoppingListItems?.set(position - 1, item)
                    navigator.onItemSelectionChange(item.isSelected)
                    notifyItemChanged(position, item)
                }


                minusDeleteCountImage.setOnClickListener {

                    val listItem = getItem(position) ?: return@setOnClickListener
                    if (enableClickEvent(listItem) || !mAdapterIsClickable) return@setOnClickListener

                    if (userShouldSetSuburb()) {
                        navigator.openSetSuburbProcess(listItem)
                        return@setOnClickListener
                    }
                    if (listItem.quantityInStock == 0) return@setOnClickListener
                    // One item selected by user in add to list
                    if (listItem.userQuantity == 1) {
                        deleteItemFromList(listItem, position)
                    } else if (listItem.userQuantity > 1) {
                        listItem.userQuantity -= 1
                        shoppingListItems!![position - 1] = listItem
                        navigator.onSubstractListItemCount(listItem)
                        notifyItemChanged(position, listItem)
                    }
                }

                addCountImage.setOnClickListener {

                    val listItem = getItem(position) ?: return@setOnClickListener
                    if (enableClickEvent(listItem) || !mAdapterIsClickable) return@setOnClickListener

                    if (userShouldSetSuburb()) {
                        navigator.openSetSuburbProcess(listItem)
                        return@setOnClickListener
                    }
                    if (listItem.quantityInStock == 0) return@setOnClickListener
                    // One item selected by user in add to list
                    if (listItem.userQuantity < listItem.quantityInStock) {
                        listItem.userQuantity += 1
                        shoppingListItems!![position - 1] = listItem
                        navigator.onAddListItemCount(listItem)
                        notifyItemChanged(position, listItem)
                    }
                }
                iconKebab.visibility = VISIBLE
                if (cbShoppingList.isChecked) {
                    iconKebab.setImageResource(R.drawable.icon_kebab_inactive)
                    iconKebab.isEnabled = false
                } else {
                    iconKebab.setImageResource(R.drawable.icon_kebab_active)
                    iconKebab.isEnabled = true
                }

                iconKebab.setOnClickListener {
                    navigator.naviagteToMoreOptionDialog(shoppingListItem)
                }
            }
        }
    }

    internal inner class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvSearchText: TextView

        init {
            tvSearchText = itemView.findViewById(R.id.textProductSearch)
        }
    }

    @Synchronized
    fun setList(listItems: ArrayList<ShoppingListItem>?) {
        if (listItems.isNullOrEmpty()) {
            return
        }
        type = getPreferredDeliveryType()
        userShouldSetSuburb = userShouldSetSuburb()
        shoppingListItems = listItems
        notifyItemRangeChanged(0, (shoppingListItems?.size ?: 0).plus(1))
        closeAllItems()
    }

    override fun getItemViewType(position: Int): Int {
        return when (position) {
            0 -> ITEM_VIEW_TYPE_HEADER
            else -> ITEM_VIEW_TYPE_BASIC
        }
    }

    fun adapterClickable(clickable: Boolean) {
        mAdapterIsClickable = clickable
    }

    private fun getItem(position: Int): ShoppingListItem? =
        shoppingListItems?.getOrNull(position - 1)

    fun userShouldSetSuburb(): Boolean = Utils.getPreferredDeliveryLocation() == null

    fun resetSelection() {
        shoppingListItems ?: return
        shoppingListItems?.filter {
            it.isSelected
        }?.map {
            it.userQuantity = 0
            it.isSelected = false
            it.isAddToCartInProgress = false
        }
        notifyItemRangeChanged(0, (shoppingListItems?.size ?: 0).plus(1))
        navigator.onItemSelectionChange(false)
    }

    fun setAddToCartProgress(itemId: String, isProgress: Boolean) {

        shoppingListItems?.filter {
            it.isAddToCartInProgress
        }?.map {
            it.userQuantity = 0
            it.isSelected = false
            it.isAddToCartInProgress = false
        }
        notifyItemRangeChanged(0, (shoppingListItems?.size ?: 0).plus(1))
    }
}